### ISSUE TITLE
Unable to enable prometheus monitoring for auto-registered Kubernetes cluster

### ISSUE THREAD

**Issue by MatthiasWinzeler** [Dynatrace/dynatrace-operator#884]
Hi folks

We were pleased to read that the helm chart v0.7.0 provides auto-registering of the kubernetes cluster. This means we can finally get rid of our custom script that sets up the kubernetes configuration after helm installation.

However, we are blocked by the following limitation, and we think plenty of other k8s customers might be too: When auto-registering, the operator does not allow to set that Prometheus Scraping should be enabled:

https://github.com/Dynatrace/dynatrace-operator/blob/v0.7.0/src/dtclient/kubernetes_settings.go#L79-L88
Since `prometheusExportersIntegrationEnabled` is `null`,  it will default to false, and thus will not be enabled:

<img width="1036" alt="image" src="https://user-images.githubusercontent.com/474567/175967933-97050dff-29a1-4665-8a30-f7f1c8f8fd56.png">

As a work around, we tried manually overwriting the config after setup:

```
curl -v ...<credentials and base url omitted>
"/config/v1/kubernetes/credentials/${cluster_id}" \
    -v \
    --header "$CONTENT_TYPE_JSON" \
    -X PUT \
    --data '{"label": "dynakube", "prometheusExportersIntegrationEnabled": true}'
...
{"error":{"code":400,"message":"Constraints violated.","constraintViolations":[{"path":"endpointUrl","message":"must not be null",
"parameterLocation":"PAYLOAD_BODY","location":null}]}}
```

However, `endpointUrl` is null, and it should stay that way. It appears as if the endpoint's validation was never adapted for the new config option which does not require `endpointUrl`.

We tried setting `endpointUrl` to a dummy value, but that doesn't work either:
```
{"error":{"code":400,"message":"Constraints violated.","constraintViolations":[{"path":null,"message":"URL must not be changed.","
parameterLocation":null,"location":null}]}}
```

So to summarize:
- It would be nice if we could express in the DynaKube CR that we want to enable the Prometheus monitoring and the Operator should respect that field on registration. We can provide a PR if that's welcome.
- The validation on the PUT credentials endpoint is faulty/outdated and might need to be aligned (but I think that's secondary for this issue, once the first point is solved)

We're on Dynatrace version 1.236, if that matters.

**Comment by floriangeigl** [comment]
thanks for your insights - I agree, that this is more towards large k8s environments, but if I'm not wrong it would also solve your use-case: If you set the defaults once with the flags, it would apply to any self-registered cluster. Defaults could obviously also be changed using the API.
out of curiosity, for scripting the Dynatrace env, are you writing your own stuff or do you use https://dynatrace-oss.github.io/dynatrace-monitoring-as-code/ ?

About the API - I think you're right - this looks like a bug. Seems the API does not like to validate self-registered AGs, as it wants an endpointurl != null. I will create a bug internally and reference this. thanks - great catch!

**Comment by MatthiasWinzeler** [comment]
@floriangeigl Thanks! Is there a way to follow that bug?

> out of curiosity, for scripting the Dynatrace env, are you writing your own stuff or do you use https://dynatrace-oss.github.io/dynatrace-monitoring-as-code/ ?

We're using a crazy homegrown ~1300 lines bash script that mostly uses curl and jq to configure things... we didn't know about Monaco, looks interesting, we'll take a look.

**Comment by floriangeigl** [comment]
@MatthiasWinzeler unfortunately, our bug tracking is not public. so there are two ways. I could try to ask my friends working on the ticket to provide news here - or you could create a ticket as a customer and mention bug "K8S-2643".

great I could at least spread some tools :) I know we use monaco internally A LOT and people love it.

**Comment by Manuel1605** [comment]
Hi @MatthiasWinzeler, 

we investigated the issue and found the following:

The main problem is, that the Configuration API ([kubernetes-credentials](https://www.dynatrace.com/support/help/dynatrace-api/configuration-api/k8s-credentials-api-api)) does not support configurations which were auto-created. 
To modify these configurations, you must use the Environment API ([Settings](https://www.dynatrace.com/support/help/dynatrace-api/environment-api/settings) and [Monitored Enitites](https://www.dynatrace.com/support/help/dynatrace-api/environment-api/entity-v2)). 

I prepared an example to show how you could achieve the wanted integration:

### Load the Monitored Entity to obtain the `entityId`. 
Therefore you need to know either the kubesystem-namespace uuid or the name specified in the dynakube.yaml:

**Request:** 
```
GET /api/v2/entities?pageSize=500&entitySelector=type(KUBERNETES_CLUSTER),kubernetesClusterId(f6e30661-5522-4962-bac4-981f82afc155)
```

**Alternative Request:** 
```
GET /api/v2/entities?pageSize=500&entitySelector=type(KUBERNETES_CLUSTER),customizedName(my-cluster)
```

**Response:** 
```
{
  "totalCount": 1,
  "pageSize": 500,
  "entities": [
    {
      "entityId": "KUBERNETES_CLUSTER-C6984369BCF99D47",
      "displayName": "my-cluster"
    }
  ]
}
```

### Use the `entityId` as `scope` to load the related Setting

**Request:**  
```
GET /api/v2/settings/objects?schemaIds=builtin:cloud.kubernetes&scopes=KUBERNETES_CLUSTER-C6984369BCF99D47
```

**Response:** 
```
{
  "items": [
    {
      "objectId": "vu9U3hXa3q0AAAABABhidWlsdGluOmNsb3VkLmt1YmVybmV0ZXMAEktVQkVSTkVURVNfQ0xVU1RFUgAQQzY5ODQzNjlCQ0Y5OUQ0NwAkN2JlZTJkM2MtNTk2Mi0zMGQ5LWFkMGEtMzE4MmRmYzE3OGMyvu9U3hXa3q0",
      "value": {
        "enabled": true,
        "label": "my-cluster",
        "clusterIdEnabled": true,
        "clusterId": "f6e30661-5522-4962-bac4-981f82afc155",
        "cloudApplicationPipelineEnabled": true,
        "openMetricsPipelineEnabled": false,
        "eventProcessingActive": false
      }
    }
  ],
  "totalCount": 1,
  "pageSize": 100
}
```
### Eventually use the `obectId` and the values returned in step 2 to update the Setting
**Request:**  
```
PUT /api/v2/settings/objects/vu9U3hXa3q0AAAABABhidWls...
```
**Body:** 
```
{
  "value": {
    "enabled": true,
    "label": "my-cluster",
    "clusterIdEnabled": true,
    "clusterId": "f6e30661-5522-4962-bac4-981f82afc155",
    "cloudApplicationPipelineEnabled": true,
    "openMetricsPipelineEnabled": true,
    "eventProcessingActive": false
  }
}
```
Please mind, that it is important that you don't work with static content but repost the `value` you received in step 2. Please only modify the `openMetricsPipelineEnabled` property.


Please note: Step 1 is not required when you update to the latest Dynatrace Version. In this case the following endpoint is available, which makes step 1 obsolete and replaces step 2: [get-objects](https://www.dynatrace.com/support/help/dynatrace-api/environment-api/settings/objects/get-objects)

**Comment by MatthiasWinzeler** [comment]
@Manuel1605 thanks for the detailed answer!

with this we could upgrade our deployments to the latest dynatrace operator and follow your workaround to enable the prom integration. 

however, do you plan to support setting `prometheusExportersIntegrationEnabled` when auto registering in the near future? If that's the case, we'd rather wait until an operator release which contains such support :)

**Comment by floriangeigl** [comment]
Hi @MatthiasWinzeler,

just for clarification: In our new API, as used in @Manuel1605 's example, `prometheusExportersIntegrationEnabled ` got renamed to `openMetricsPipelineEnabled `. So with the new API you could auto-register the cluster and then use your bash script to enable Prometheus scraping by setting `openMetricsPipelineEnabled ` to true. I hope this helps for now :)

Currently, we don't plan to make `openMetricsPipelineEnabled ` (former `prometheusExportersIntegrationEnabled `) part of the DynaKube. Instead, we plan to implement the approach with the hierarchical settings as sketched in my previous [reply](https://github.com/Dynatrace/dynatrace-operator/issues/884#issuecomment-1168776760).

As there is a way to achieve what was asked in this issue, I would love to close it, if this is okay for you @MatthiasWinzeler ?

**Comment by MatthiasWinzeler** [comment]
@floriangeigl Sure, makes sense! We'll then go forward and implement the proposed workaround for the time being.

Thanks for your help; I'll close it.
